tests/test_files.py

```diff
@@ -1,223 +1,89 @@
-
-    assert not is_ignored("dir")
-    assert not is_ignored("dir/link")
-    assert not is_ignored("dir/link/link")
-
-    assert is_ignored("dir/ignored")
-    assert is_ignored("dir/link/ignored")
-    assert is_ignored("dir/link/link/ignored")
-
-
-def test_ignore_git_symlink_outside_repo(
-    tmp_path: pathlib.Path, monkeypatch: pytest.MonkeyPatch
-) -> None:
-    monkeypatch.chdir(tmp_path)
-
-    (tmp_path / "repo" / ".git").mkdir(parents=True)
-    (tmp_path / "repo" / ".gitignore").write_text("link")
-
-    (tmp_path / "link").mkdir()
-    (tmp_path / "repo" / "link").symlink_to(tmp_path / "link")
-
-    assert not is_ignored("link")
-    assert not is_ignored("link/main.py")
-    assert is_ignored("repo/link")
-    assert is_ignored("repo/link/main.py")
-
-
-def test_ignore_symlink_circular(
-    tmp_path: pathlib.Path, monkeypatch: pytest.MonkeyPatch
-) -> None:
-    monkeypatch.chdir(tmp_path)
-
-    (tmp_path / "link1").symlink_to(tmp_path / "link2")
-    (tmp_path / "link2").symlink_to(tmp_path / "link1")
-
-    assert not is_ignored("link1")
-    assert not is_ignored("link2")
+from ssort._files import current_working_dir, find_project_root
+
+
+def test_current_working_dir():
+    assert current_working_dir() == Path(".").resolve()
+
+
+class TestFindProjectRoot:
+    @pytest.fixture(
+        params=[
+            (".",),
+            (".", "dir"),
+            ("dir", "dirA/B"),
+            ("dirA/B", "dir"),
+            ("a/a/a", "b/b/b"),
+            ("dir/a", "dir/b"),
+        ]
+    )
+    def subdir(self, request):
+        return request.param
+
+    @pytest.fixture()
+    def mock_current_working_dir(self, mocker, tmp_path):
+        return mocker.patch(
+            "ssort._files.current_working_dir", return_value=tmp_path / "root"
+        )
+
+    @pytest.fixture()
+    def git(self, tmp_path):
+        root = tmp_path / "root"
+        (root / ".git").mkdir(parents=True)
+        return root
+
+    def test_find_project_root_git(
+        self, subdir, git, mock_current_working_dir
+    ):
+        print(subdir)
+        patterns = [git / sub for sub in subdir]
+
+        for p in patterns:
+            p.mkdir(parents=True, exist_ok=True)
+
+        assert git == find_project_root(patterns)
+        mock_current_working_dir.assert_called_once()
+
+    @pytest.fixture()
+    def pyproject(self, tmp_path):
+        root = tmp_path / "root"
+        root.mkdir()
+        (root / "pyproject.toml").touch()
+        return root
+
+    def test_find_project_root_pyproject(
+        self, subdir, pyproject, mock_current_working_dir
+    ):
+        patterns = [pyproject / sub for sub in subdir]
+
+        for p in patterns:
+            p.mkdir(parents=True, exist_ok=True)
+
+        assert pyproject == find_project_root(patterns)
+        mock_current_working_dir.assert_called_once()
+
+    @pytest.fixture()
+    def neither(self, tmp_path):
+        root = tmp_path / "root"
+        root.mkdir()
+        return root
```

Comment:
I don't think that having this code in fixtures is helpful.  At best, it adds a layer of indirection with no abstraction.  At worst, it encourages coupling.  A variable binding provides equivalent abstraction with no indirection.

Reply:
I see where the confusion is!  I was referring specifically to the `pytest.fixture` when I said fixture.  I agree that these are _fixtures_, but that doesn't mean that they have to be `pytest.fixture`s.

I would argue that there are at least two sorts of _fixtures_:
  1) Fixtures which provide the background in which your test runs:  e.g. `tmp_path` or references to a default-initialised database.
  2) Fixtures which set the specific state against which you want to test:  e.g. specific files with in a temporary directory, or particular rows in a database.
  
There are also other things which use `pytest.fixture` but that aren't fixtures:  `pytest.monkeypatch` is a good example as it just uses `pytest.fixture` for lifecycle management but doesn't set up any state.

Type 1 fixtures are just background and should be attract as little attention as possible when added to a test.  They should usually be defined outside the test and can often be shared.

Type 2 fixtures are core to understanding what a test is asserting and so it makes sense for them to be inline and explicit.  Changing them changes what all tests that use them assert so sharing them is dangerous but it may be possible to share some code to construct them.

To my mind, these are definitely type 2 fixtures and so should be inline.